File: LeetCode_Solutions/LeetCode_Solutions/ReverseLinkedList_206.m
//
//  ReverseLinkedList_206.m
//  LeetCode_Solutions
//
//  Created by 来自遥远星系的核心巡洋舰 on 16/10/28.
//  Copyright © 2016年 KnightJoker. All rights reserved.
//


#import "ReverseLinkedList_206.h"

@implementation ReverseLinkedList_206

- (struct ListNode *)reverseList:(struct ListNode *)head {

    if (!head || !head -> next) {
        return head;
    }
    
    struct ListNode *nextP = head -> next;
    struct ListNode *resP = [self reverseList:head->next];
    
    nextP -> next = head;
    head -> next = nil;
    return resP;
}

@end

File: LeetCode_Swift/LeetCode_Swift/ReverseLinkedList_206.swift
//
//  ReverseLinkedList_206.swift
//  LeetCode_Swift
//
//  Created by Huni on 16/08/2017.
//  Copyright © 2017 KnighhtJoker. All rights reserved.
//

import Foundation

public class ReverseLinkedList_206 {

    func reverseList(_ head: ListNode?) -> ListNode? {
        if head == nil || head?.next == nil {
            return head
        }
        
        let next = head?.next
        let res = self.reverseList(head?.next)
        
        next?.next = head
        head?.next = nil
        return res
        
    }
}
